### PR TITLE
Change the proto to support multiple network architectures.

### PR DESCRIPTION
**Pull request by mooskagh**

### CODE REVIEW

File: proto/net.proto [hunk 1]
```diff
@@ -81,14 +81,45 @@ message TrainingParams {
...
     UNKNOWN = 0;
     LINEAR16 = 1;
-    ZFP = 2;
   }
 
-  optional Encoding weights_encoding = 1;
+  // Deprecated and will be removed in future versions.
+  // Please do not populate this field if you set network_format
+  // (for engine v0.20 and later).
+  optional Encoding weights_encoding = 1 [deprecated = true];
```
Comment (Tilps) [inline proto/net.proto]: Encoding and network format should be independent axes - so I don't think we should deprecate weights_encoding unless you want to make Encoding a member of NetworkFormat. (Which it currently is not.)

File: proto/net.proto [hunk 2]
```diff
@@ -81,14 +81,45 @@ message TrainingParams {
...
+    INPUT_UNKNOWN = 0;
+    INPUT_CLASSICAL_112_PLANE = 1;
+  }
+  optional InputFormat input = 1;
+
+  // Output format of the NN. Used by search code to interpret results.
+  enum OutputFormat {
+    OUTPUT_UNKNOWN = 0;
+    OUTPUT_CLASSICAL = 1;
+    // OUTPUT_WDL = 2;  // Uncomment when implemented.
```
Comment (Tilps) [inline proto/net.proto]: There are two kinds of output - value head and policy head.  This commented enum value suggests that this enum is actually specifically for the value head axis.  So maybe it should be called ValueOutputFormat and value_output.

(That being said given the network structure it should be obvious (from the number of bytes in the last layer of the value head) whether its a 3 valued output or a 1 valued output so this enum could be avoided entirely unless we plan on having some very strange value head options in future.)

File: proto/net.proto [hunk 3]
```diff
@@ -81,14 +81,45 @@ message TrainingParams {
...
+  }
+  optional OutputFormat output = 2;
+
+  // Network architecture. Used by backends to build the network.
+  enum NetworkStructure {
+    NETWORK_UNKNOWN = 0;
+    NETWORK_CLASSICAL = 1;
+    // NETWORK_SE = 2;  // Uncomment when implemented.
+  }
+  optional OutputFormat network = 3;
```
Comment (Tilps) [inline proto/net.proto]: I assume this should be NetworkStructure enum.

File: proto/net.proto (hunk 1 above)
Comment (mooskagh) [reply proto/net.proto]: Keeping it but not actually use in the code for now. Will checkfail if it contains anything other than LINEAR16.

File: proto/net.proto (hunk 3 above)
Comment (mooskagh) [reply proto/net.proto]: Done.

File: proto/net.proto (hunk 2 above)
Comment (mooskagh) [reply proto/net.proto]: As discussed, will keep just one enum for output, but will change GetPVal and GetQVal API to just be GetOutputVal()

**Pull request merged**